src/core/filesystem.cc

```diff
@@ -1150,19 +1151,23 @@ Status
 S3FileSystem::ParsePath(
     const std::string& path, std::string* bucket, std::string* object)
 {
+  // Cleanup trailing and leading slashes
+  std::string clean_path;
+  RETURN_IF_ERROR(CleanPath(path, &clean_path));
+
   // Get the bucket name and the object path. Return error if input is malformed
   std::string host_name, host_port;
   if (!RE2::FullMatch(
-          path, s3_regex_, &host_name, &host_port, bucket, object)) {
-    int bucket_start = path.find("s3://") + strlen("s3://");
-    int bucket_end = path.find("/", bucket_start);
+          clean_path, s3_regex_, &host_name, &host_port, bucket, object)) {
+    int bucket_start = clean_path.find("s3://") + strlen("s3://");
```

Comment:
Isn't the path required to start with exactly "s3://"? Can't you just check this once at the beginning of the function and return error if fails, and thereafter just work with the rest of the path. That should also make it much easier to remove the leading and trainling /.